packages/flutter_tools/lib/src/macos/build_macos.dart

```diff
@@ -140,46 +140,62 @@ Future<void> buildMacOS({
   if (result != 0) {
     throwToolExit('Build process failed');
   }
-  if (buildInfo.codeSizeDirectory != null && sizeAnalyzer != null) {
-    final String arch = DarwinArch.x86_64.name;
-    final File aotSnapshot = globals.fs.directory(buildInfo.codeSizeDirectory)
-      .childFile('snapshot.$arch.json');
-    final File precompilerTrace = globals.fs.directory(buildInfo.codeSizeDirectory)
-      .childFile('trace.$arch.json');
 
-    // This analysis is only supported for release builds.
-    // Attempt to guess the correct .app by picking the first one.
-    final Directory candidateDirectory = globals.fs.directory(
-      globals.fs.path.join(getMacOSBuildDirectory(), 'Build', 'Products', 'Release'),
-    );
-    final Directory appDirectory = candidateDirectory.listSync()
-      .whereType<Directory>()
-      .firstWhere((Directory directory) {
-      return globals.fs.path.extension(directory.path) == '.app';
-    });
-    final Map<String, Object?> output = await sizeAnalyzer.analyzeAotSnapshot(
-      aotSnapshot: aotSnapshot,
-      precompilerTrace: precompilerTrace,
-      outputDirectory: appDirectory,
-      type: 'macos',
-      excludePath: 'Versions', // Avoid double counting caused by symlinks
-    );
-    final File outputFile = globals.fsUtils.getUniqueFile(
-      globals.fs
-        .directory(globals.fsUtils.homeDirPath)
-        .childDirectory('.flutter-devtools'), 'macos-code-size-analysis', 'json',
-    )..writeAsStringSync(jsonEncode(output));
-    // This message is used as a sentinel in analyze_apk_size_test.dart
-    globals.printStatus(
-      'A summary of your macOS bundle analysis can be found at: ${outputFile.path}',
-    );
-
-    // DevTools expects a file path relative to the .flutter-devtools/ dir.
-    final String relativeAppSizePath = outputFile.path.split('.flutter-devtools/').last.trim();
-    globals.printStatus(
-      '\nTo analyze your app size in Dart DevTools, run the following command:\n'
-      'dart devtools --appSizeBase=$relativeAppSizePath'
-    );
+  if (_shouldWriteCodeSizeAnalysis(buildInfo, sizeAnalyzer)) {
+    await _writeCodeSizeAnalysis(buildInfo, sizeAnalyzer!);
```

Comment:
The `!` is a bummer. What do you think of unconditionally calling `_writeCodeSizeAnalysis` but having `_writeCodeSizeAnalysis` no-op if it shouldn't run? This would allow us to remove the `!`. I don't feel strongly about this, feel free to skip.

Reply:
Inlined the check; it makes the comment a bit wordier, but avoids the `!`.